pattern.c

```diff
@@ -1549,67 +1556,72 @@ struct Pattern *mutt_pattern_comp(/* const */ char *s, int flags, struct Buffer
           if (!*ps.dptr)
           {
             mutt_buffer_printf(err, "%s", _("missing parameter"));
-            mutt_pattern_free(&curlist);
-            return NULL;
+            goto cleanup;
           }
-          if (!entry->eat_arg(tmp, &ps, err))
+          if (!entry->eat_arg(pat, &ps, err))
           {
-            mutt_pattern_free(&curlist);
-            return NULL;
+            goto cleanup;
           }
         }
         implicit = true;
         break;
+      }
+
       case '(':
+      {
         p = find_matching_paren(ps.dptr + 1);
         if (*p != ')')
         {
           mutt_buffer_printf(err, _("mismatched parenthesis: %s"), ps.dptr);
-          mutt_pattern_free(&curlist);
-          return NULL;
+          goto cleanup;
         }
         /* compile the sub-expression */
         buf = mutt_str_substr_dup(ps.dptr + 1, p);
         tmp = mutt_pattern_comp(buf, flags, err);
-        if (!tmp)
-        {
-          FREE(&buf);
-          mutt_pattern_free(&curlist);
-          return NULL;
-        }
         FREE(&buf);
+        if (!tmp)
+          goto cleanup;
+        struct Pattern *pat = SLIST_FIRST(tmp);
         if (last)
-          last->next = tmp;
+          SLIST_NEXT(SLIST_FIRST(last), entries) = pat;
         else
           curlist = tmp;
         last = tmp;
-        tmp->not ^= not;
-        tmp->alladdr |= alladdr;
-        tmp->isalias |= isalias;
+        pat = SLIST_FIRST(tmp);
+        pat->not ^= not;
+        pat->alladdr |= alladdr;
+        pat->isalias |= isalias;
         not = false;
         alladdr = false;
         isalias = false;
         ps.dptr = p + 1; /* restore location */
         break;
+      }
+
       default:
         mutt_buffer_printf(err, _("error in pattern at: %s"), ps.dptr);
-        mutt_pattern_free(&curlist);
-        return NULL;
+        goto cleanup;
     }
   }
   if (!curlist)
```

Comment:
`curlist` can never be an empty list since `mutt_pattern_new_node` (which is probably inappropriately named) always creates a new list with a new node and that's the only way mutt_pattern_comp internally creates a new list.

Do you think this is worth clarifying as a comment in the function?

Reply:
> `mutt_pattern_new_node` (which is probably inappropriately named)

Yeah, it's a bit of an odd function.
Add a comment for now and we'll think about it.

Reply:
Added comment to clarify curlist is always non-empty.